clis/teliod/src/main.rs

```diff
@@ -129,19 +194,169 @@ impl CommandListener {
+                // exit daemon
+                return Err(TeliodError::UpdateMachineError(status));
+            }
+            StatusCode::NOT_FOUND => {
+                // Retry machine update after registering. To make sure everything was successful
+                // If registering fails. Close the daemon
+                client_config.machine_identifier = register_machine(
+                    &client_config.hw_identifier,
+                    client_config.public_key,
+                    &client_config.auth_token,
+                )
+                .await?;
+            }
+            _ => {
+                // Retry with exp back-off
+                if retries < MAX_RETRIES {
+                    let backoff_time = BASE_DELAY_MS * 2_u64.pow(retries);
+                    let sleep_time = min(backoff_time, MAX_BACKOFF_TIME);
+                    debug!("Retrying after {} ms...", sleep_time);
+                    tokio::time::sleep(Duration::from_millis(sleep_time)).await;
+                    retries += 1;
+                } else {
+                    // If max retries exceeded, exit daemon
+                    return Err(TeliodError::UpdateMachineTimeoutError);
+                    // panic!("Max retries reached. Exiting.");
+                }
+            }
+        }
+    }
+
+    config.private_key = Some(client_config.private_key.clone());
+    config.hw_identifier = Some(client_config.hw_identifier.clone());
+    config.machine_identifier = Some(client_config.machine_identifier.clone());
+
+    let mut file = File::create(config_path)?;
+    serde_json::to_writer(&mut file, &config)?;
+    Ok(())
+}
+
 // From async context Telio needs to be run in separate task
-fn telio_task() {
-    let _telio = Device::new(
+fn telio_task(
+    client_config: ClientConfig,
+    mut rx_channel: mpsc::Receiver<TelioTaskCmd>,
+) -> Result<(), TeliodError> {
+    debug!("Initializing telio device");
+    let mut telio = Device::new(
         Features::default(),
         // TODO: replace this with some real event handling
         move |event| info!("Incoming event: {:?}", event),
         None,
-    );
+    )?;
+
+    // TODO: This is temporary to be removed later on when we have proper integration
+    // tests with core API. This is to not look for tokens in a test environment
+    // right now as the values are dummy and program will not run as it expects
+    // real tokens.
+    if !client_config.auth_token.eq("") {
+        start_telio(&mut telio, client_config.private_key)?;
+        if let Err(e) = update_meshmap(&client_config, &telio) {
+            error!("Unable to set meshmap due to {e}");
+        }
+
+        while let Some(cmd) = rx_channel.blocking_recv() {
+            info!("Got command {:?}", cmd);
+            match cmd {
+                TelioTaskCmd::UpdateMeshmap => {
+                    if let Err(e) = update_meshmap(&client_config, &telio) {
```

Comment:
As I said before, I think Telio task should block just on the commands and Telio API, but maybe it's not that but, I don't insist on moving it outside (but I think it would improve the quality of this code)

Reply:
If we use async to get_meshmap we would still have to use block_recv to get the meshmap part, which would in the end come back to the same problem wouldn't it ?

Reply:
Telio task would wait on the whole channel and could just use the received meshmap when it comes, so if any other thing  comes from the channel when meshmap is being downloaded, it could just use it